File: examples/javascript/valid.js
#!/usr/bin/env node
/**
 * A valid JavaScript example implementing a data processing utility.
 * This file should pass validation with Synx.
 */

// Data processing class with modern JS features
class DataProcessor {
  constructor(data = []) {
    this.data = data;
    this.errors = [];
  }

  // Process array of numbers with error handling
  processNumbers() {
    try {
      // Check if we have data
      if (!this.data.length) {
        throw new Error('No data to process');
      }

      // Array methods and arrow functions
      const results = {
        sum: this.data.reduce((acc, val) => acc + val, 0),
        average: this.data.reduce((acc, val) => acc + val, 0) / this.data.length,
        min: Math.min(...this.data),
        max: Math.max(...this.data),
        median: this.#calculateMedian(),
        processed: this.data.map(x => ({
          original: x,
          doubled: x * 2,
          squared: x ** 2
        }))
      };

      return results;
    } catch (error) {
      this.errors.push(error.message);
      return null;
    }
  }

  // Private method (using modern JS feature)
  #calculateMedian() {
    const sorted = [...this.data].sort((a, b) => a - b);
    const middle = Math.floor(sorted.length / 2);
    
    if (sorted.length % 2 === 0) {
      return (sorted[middle - 1] + sorted[middle]) / 2;
    } else {
      return sorted[middle];
    }
  }

  // Generator function to yield processed items one by one
  *processedItemsGenerator() {
    for (const item of this.data) {
      yield {
        original: item,
        squared: item ** 2
      };
    }
  }
}

// Example usage
const numbers = [12, 5, 8, 21, 16, 7, 3, 9];
const processor = new DataProcessor(numbers);
const results = processor.processNumbers();

console.log('Processing results:', results);

// Using the generator
console.log('\nProcessed items one by one:');
for (const item of processor.processedItemsGenerator()) {
  console.log(item);
}

File: examples/python/valid.py
#!/usr/bin/env python3
"""
A valid Python example that calculates Fibonacci numbers.
This file should pass validation with Synx.
"""

def fibonacci(n):
    """Calculate the nth Fibonacci number recursively with memoization."""
    memo = {0: 0, 1: 1}
    
    def fib_memo(n):
        if n not in memo:
            memo[n] = fib_memo(n-1) + fib_memo(n-2)
        return memo[n]
    
    return fib_memo(n)

def is_prime(num):
    """Check if a number is prime."""
    if num < 2:
        return False
    for i in range(2, int(num**0.5) + 1):
        if num % i == 0:
            return False
    return True

if __name__ == "__main__":
    print("Fibonacci Sequence:")
    for i in range(10):
        print(f"fibonacci({i}) = {fibonacci(i)}")
    
    print("\nPrime Fibonacci Numbers:")
    for i in range(20):
        fib = fibonacci(i)
        if is_prime(fib):
            print(f"fibonacci({i}) = {fib} is prime")

